File: Problems/Algorithms/841. Keys and Rooms/KeysRooms.java
class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        int n = rooms.size();
        boolean[] visited = new boolean[n];
        visited[0] = true;
        
        Stack<Integer> stack = new Stack<>();
        stack.push(0);
        
        while (!stack.isEmpty()) {
            int curr = stack.pop();
            
            for (int neighbor : rooms.get(curr)) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    stack.add(neighbor);
                }
            }
        }
        
        for (boolean v : visited) {
            if (!v) return false;
        }
        return true;
    }
}

File: Problems/Algorithms/841. Keys and Rooms/KeysRooms.kt
class Solution {
    fun canVisitAllRooms(rooms: List<List<Int>>): Boolean {
        val n = rooms.size
        val visited = BooleanArray(n) { false }
        visited[0] = true
        
        val stack = ArrayDeque<Int>()
        stack.add(0)
        
        while (!stack.isEmpty()) {
            val curr = stack.removeLast()
            
            for (neighbor in rooms.get(curr)) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true
                    stack.add(neighbor)
                }
            }
        }
        
        for (v in visited) {
            if (!v) return false
        }
        return true
    }
}

